TASK: Rewrite Neos.Neos:Shortcut to pure fusion

mhsdesign opened the pull request:
The backend rendering for shortcut nodes is reimplemented in the fusion prototype `Neos.Neos:Shortcut` and the previous fluid template is removed.

resolves: #3261


**I tested all possible backend views of the Neos.Neos:Shortcut in the Neos.Demo:**
- [x] shortcut to the parent page
- [x] shortcut to image
- [x] shortcut to external URI
- [x] shortcut to specific node target / reference
- [x] shortcut to first child page

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,132 @@
+    entryIdentifier {
+      node = ${node}
+		}
+    entryDiscriminator = ${q(Neos.BackendUser.get()).property('preferences.interfaceLanguage')}
```
as explained here: https://github.com/neos/neos-development-collection/pull/3372#issue-682825113

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,132 @@
+        fallback = afx`
+          Click {this.targetUriTag} to continue to the page.
+        `
+        renderer = ${I18n.id('shortcut.clickToContinueToPage').package('Neos.Neos').source('Main').arguments([this.targetUriTag]).value(this.fallback).locale(props.i18n).translate()}
```
Yeah these lines are a bit long and mostly the same (besides the id). They use `this.targetUriTag` and `this.fallback`.

Should this be implemented a bit cleaner? With a custom EEL helper f.x.?

mhsdesign replied:
Solved thanks to https://mind-the-seb.de/de/blog/der-translation-helper-in-neos-cms

daniellienert commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,133 @@
+    @subject = ${props.targetMode}
+  }
+
+  // targetMode == 'selectedTarget'
```
This fusion is pretty hard to read. I would generally recommend:

- merge and nest the conditions
- remove the fallbacks and add them to the english language file so that they are translateable

mhsdesign replied:
okay thanks ^^
about the language files - the english language file contains most of the sources except for 'shortcut.toSpecificTarget' im sure youre editing the files not per hand - how would i add 'shortcut.toSpecificTarget' ?

daniellienert commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,133 @@
+      }
+
+      asset = Neos.Fusion:Renderer {
+        targetUriTag = Neos.Fusion:Tag {
```
Use afx for the tag would save several lines and thus make it more readable

daniellienert commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,133 @@
+    </div>
+  `
+
+  @cache {
```
Why is the caching configuration needed when it was not once it was a template?

mhsdesign replied:
i just found it out while testing, and it annoyed me:
when you have visited a Shortcut in the backend and then change the user language, the cache will still show you the previous translation in the backend.

... i now since this is hardly ever the case i could also remove it right?

daniellienert commented on Neos.Neos/Classes/Fusion/Helper/BackendUserHelper.php:
```diff
@@ -0,0 +1,45 @@
+/**
+ * get the current user from the backend
+ */
+class BackendUserHelper implements ProtectedContextAwareInterface {
```
It may be better to add a specific BackendTranslationHelper instead of exposing the backend user completely to fusion.

daniellienert reviewed (dismissed):
Hey @mhsdesign,

thank you very much for working on that task!!!

I think we can shrink the fusion down and make it more readable, see my comments. Currently fusion only renders in frontend language, so we should discuss, if we should add a backendTranslationHelper instead of a BackendUserHelper.

mhsdesign reviewed (commented):
okay thanks for your advice ;)

i think a BackendTranslationHelper might be better too. I will work on it ;)

mhsdesign commented:
Some points left - wich need to be discussed:

* i found some css in https://github.com/neos/neos-ui/blob/master/Resources/Private/Fusion/Prototypes/Shortcut.fusion
which actually belongs to this prototype ... should i move it here then?

* the 

... btw i will remove the translation fallbacks and add the missing 'shortcut.toSpecificTarget'

mhsdesign commented:
Some points left - wich need to be discussed:

- [x]  i found some css in https://github.com/neos/neos-ui/blob/master/Resources/Private/Fusion/Prototypes/Shortcut.fusion
which actually belongs to this prototype ... should i move it here then? yes -> but how without breaking  things... https://github.com/neos/neos-ui/commit/ec838d9f521890ecff4d4f9ef5fcfbac218e9fb8 -> also inline style cheets - i was told - is not THE way

- [x] i disabled the cache for the backend shortcut view - as it previously didnt calculate the current user language into the cache, and i figured i wouldnt cache backend modules either.

- [x] i will remove the fallback translations soon but the 'shortcut.toSpecificTarget' (see #3371) translation is missing. I read that Neos uses Crowdin - so can someone with access please handle that 😁 

- [x] ~i dont know if the BackendTranslationHelper is well written, as it extends the flow translation helper and adjusts a protected method. But this was the cleanest? way to support shorthand translation, fullparameter translation and chained translation. Otherwise one need to copy paste nearly half of the flow translation helper~  solved by https://github.com/neos/neos-development-collection/pull/3939

mhsdesign commented:
oops... crowdin is no more? https://github.com/neos/neos-development-collection/pull/3437#issuecomment-921705711

mhsdesign commented:
~since im extending the flow translation helper and modifing a protecteded method - this will fail silently if the flow guys rename the method. Should i implement a method_exists check in the constructor? I guess the best would be to refactor the flow translation helper to make it easily extensible.~

mhsdesign commented:
~So i thought about this again, the fluid `<neos:backend.translate>` implementation:
https://github.com/neos/neos-development-collection/blob/ddf0b3f273395cb4dee316c5d0205b603abba5f7/Neos.Neos/Classes/ViewHelpers/Backend/TranslateViewHelper.php uses also inheritance and some ugly duplicated code ... id like to avoid that, but there is no simple way around that.~

~Similar to the fluid view helper, it would look for the eel helper like this (by @jonnitto ):
https://github.com/CarbonPackages/Carbon.Eel/blob/b2cf962d4dc367f79d41d573cc9766223ed0787a/Classes/EelHelper/BackendHelper.php~

~I just see inheritance in this case as not a good design choice.~

mhsdesign commented:
this is WIP, until https://github.com/neos/neos-development-collection/issues/3638 is solved.

mhsdesign committed:
TASK Rewrite Neos.Neos:Shortcut to pure Fusion

mhsdesign committed:
TASK Neos.Neos:Shortcut remove translation fallbacks

with https://github.com/neos/neos-development-collection/pull/3937 we dont need the fallbacks anymore ;)

mhsdesign committed:
TASK Neos.Neos:Shortcut cleanup Fusion nesting

mhsdesign committed:
TASK Neos.Neos:Shortcut dont cache the backend view

Otherwise, changes from `Neos.Backend.interfaceLanguage()` will not be reflected.

And its like a backend module -> which i wouldn't cache also.

mhsdesign commented:
With https://github.com/neos/neos-development-collection/pull/3937 (waiting for upmerge) and https://github.com/neos/neos-development-collection/issues/3371 in place, the path of my first Neos PR can continue after beeing 14 months stale! ;)

Open for reviews ;)

mhsdesign committed:
Merge remote-tracking branch 'origin/8.2' into FusionRewrite-NeosNeosShortcut

mhsdesign commented:
Okay, #3937 is now upmerged (and available in this branch), so now it's really ready for review ;)

mficzel reviewed (changes requested):
Works as expected and and removes the fluid template.

1. The pr should start with a brief description that is suitable for the release notes like:

> The backend rendering for shortcut nodes was is reimplemented via the new fusion prototype `Neos.Neos:Shortcut` and the previous fluid template is removed.

2. While this is pretty much an 1:1 fusion implementation. i personally would prefer to use some prototypes inside to make this extensible but consider this optional. 
It could work with a central `Neos.Neos:Shortcut.Case` that does the switching between `Neos.Neos:Shortcut.ParentNode`, `Neos.Neos:Shortcut.FirstChildNode` and `Neos.Neos:Shortcut.SelectedTargetNode`

jonnitto commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,83 @@
+        }
+
+        asset = Neos.Fusion:Value {
+          targetUriTag = afx`
+            <a target='_blank'>
+              <Neos.Fusion:ResourceUri @path='attributes.href' resource={props.targetConverted.resource} />
+              {props.targetConverted.label}
+            </a>
+          `
```
I would prefer pure Fusion here…

mhsdesign replied:
is adjusted ;)

Funnily i had it that way first https://github.com/neos/neos-development-collection/pull/3372#discussion_r663644088

mhsdesign committed:
BUGFIX 3835 apply suggestions from code review

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/Prototypes/Shortcut.fusion:
```diff
@@ -1,15 +1,87 @@
+
+    parentNode = Neos.Fusion:Value {
+      targetUriTag = Neos.Neos:NodeLink {
+        node = ${props.node.parent}
```
upsi that cannot work ;)

need to use `node.parent` directly

bwaidelich replied:
still to fix?

mhsdesign commented:
> 2\. While this is pretty much an 1:1 fusion implementation. i personally would prefer to use some prototypes inside to make this extensible but consider this optional.
>     It could work with a central `Neos.Neos:Shortcut.Case` that does the switching between `Neos.Neos:Shortcut.ParentNode`, `Neos.Neos:Shortcut.FirstChildNode` and `Neos.Neos:Shortcut.SelectedTargetNode`

good idea (WIP at my pc) but ill wait for https://github.com/neos/neos-development-collection/pull/3943 ;)

also those parts need to be discussed https://github.com/neos/neos-development-collection/pull/3372#issuecomment-922064157

bwaidelich commented:
Thanks for this!

I agree to @mficzel's comments. Splitting this into smaller chunks would greatly enhance extensibility and readability.

> good idea (WIP at my pc) but ill wait for https://github.com/neos/neos-development-collection/pull/3943 ;)

Can we maybe get this in based on the _current_ feature set? IMO #3943 is a great direction but it's a higher-risk-feature and it would be a pitty if it blocked this one.

Regarding

> i found some css [...] should i move it here then? yes -> but how without breaking things

What about targeting 9.0 with this feature?

> i disabled the cache for the backend shortcut view - as it previously didnt
> calculate the current user language into the cache

Can't we add the user language to the `@cache` config?

bwaidelich commented:
It's been a while.. Is this one still in progress?

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into FusionRewrite-NeosNeosShortcut

mhsdesign committed:
TASK: Adjust `Neos.Neos:Shortcut` to work with Neos 9

mhsdesign committed:
TASK: Rename Neos.scss Main.scss to be consistent

The compiled file is also named Main.css

mhsdesign committed:
TASK: Move Shortcut.css from Ui to Neos

mhsdesign commented:
Neos ui part https://github.com/neos/neos-ui/pull/3657

mhsdesign commented on Neos.Neos/Resources/Private/Styles/Main.scss:
This was incorrectly named `Neos` and should be `Main`... lead to confusion. The compiled file is named `Main.css`

mficzel reviewed (approved):
Looks good by reading and it works
